DOC Newly supported and deprecated assets API (#107)
silverstripe/developer-docs: en/04_Changelogs/4.12.0.md

See [managing records](/developer_guides/model/data_model_and_orm/managing_records/) for a detailed explanation of this extension.

### Assets API formalisation and deprecations {#assets-api-formalisation-deprecation}

> In the [Silverstripe CMS 4.4 release](4.4.0/#optional-migration-tasks-overview), many aspects of the `silverstripe/assets` module were refactored to support permanent link.

Reviewer: ```suggestion
In the [Silverstripe CMS 4.4 release](4.4.0/#optional-migration-tasks-overview), many aspects of the `silverstripe/assets` module were refactored to support permanent links.
```

Many new abstractions were created in the process. However, those new abstraction were marked as `@internal` and were not covered by Silverstripe CMS's semantic versioning commitment. Multiple file migration tasks were also added in that releases.

> With Silverstripe CMS 5 on the horizon, it is an opportune time to formalise some of those abstractions and to deprecate those migration task.

Reviewer: ```suggestion
With Silverstripe CMS 5 on the horizon, it is an opportune time to formalise some of those abstractions and to deprecate those migration tasks.
```

- [`SilverStripe\Assets\Storage\Sha1FileHashingService`](api:SilverStripe\Assets\Storage\Sha1FileHashingService)
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::getPublicResolutionStrategy()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::getPublicResolutionStrategy())
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::setPublicResolutionStrategy()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::setPublicResolutionStrategy())
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::getProtectedResolutionStrategy()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::getProtectedResolutionStrategy())
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::setProtectedResolutionStrategy()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::setProtectedResolutionStrategy())
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::applyToFileOnFilesystem()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::applyToFileOnFilesystem())
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::applyToFileIDOnFilesystem()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::applyToFileIDOnFilesystem())

> These new abstraction, make several APIs redundant. The following method are deprecated in the Silverstripe CMS 4.12 release and will be removed in Silverstripe CMS 5:

Reviewer: ```suggestion
These new abstractions make several APIs redundant. The following methods are deprecated in the Silverstripe CMS 4.12 release and will be removed in Silverstripe CMS 5:
```

- [`SilverStripe\Assets\FilenameParsing\LegacyFileIDHelper`](api:SilverStripe\Assets\FilenameParsing\LegacyFileIDHelper)
- [`SilverStripe\Assets\Flysytem\FlysystemAssetStore::findVariants()`](api:SilverStripe\Assets\Flysytem\FlysystemAssetStore::findVariants())

> Silverstripe CMS 4 includes multiple task to migrate files to newer states. These task will no longer be needed in Silverstripe CMS 5. The following task and helpers have been deprecated in Silverstripe CMS 4 and will be removed in Silverstripe CMS 5:

Reviewer: ```suggestion
Silverstripe CMS 4 includes multiple tasks to migrate files to newer states. These tasks will no longer be needed in Silverstripe CMS 5. The following tasks and helpers have been deprecated in Silverstripe CMS 4 and will be removed in Silverstripe CMS 5:
```
